File: BlotoutAnalytics/BlotoutAnalyticsAll/Network/BONetworkManager.m
//
//  BONetworkManager.m
//  BlotoutAnalytics
//
//  Created by Blotout on 05/09/19.
//  Copyright © 2019 Blotout. All rights reserved.
//

#import "BONetworkManager.h"
#import <BlotoutFoundation/BOFNetworkPromise.h>
#import <BlotoutFoundation/BOFNetworkPromiseExecutor.h>
#import <BlotoutFoundation/BOReachability.h>
#import "NSError+BOAdditions.h"
#import <BlotoutFoundation/BOFLogs.h>
#import "BOANetworkConstants.h"

@implementation BONetworkManager


+ (void)asyncRequest:(NSURLRequest *)request
             success:(void(^)(id, NSURLResponse *))successBlock_
             failure:(void(^)(id, NSURLResponse *, NSError *))failureBlock_ {
    
    @try {
        if ([[BOReachability reachabilityForInternetConnection] currentReachabilityStatus] == BONotReachable) {
            NSError *error = [NSError boErrorForCode:BOErrorNoInternetConnection withMessage:@"Network Not Reachable"];
            failureBlock_(nil, nil,error);
            return;
        }
        
        BOFNetworkPromise *netpromise = [[BOFNetworkPromise alloc] initWithURLRequest:request completionHandler:^(NSURLResponse * _Nullable urlResponse, id  _Nullable dataOrLocation, NSError * _Nullable error) {
            
            NSHTTPURLResponse *httpRes = (NSHTTPURLResponse*)urlResponse;
            
            if(httpRes.statusCode  == 200) {
                successBlock_(dataOrLocation,urlResponse);
            } else {
                failureBlock_(dataOrLocation,urlResponse,error);
            }
            
            BOFLogDebug(@"urlResponse:%@  dataOrLocation:%@ error:%@ urlResponse_statusCode:%ld allHeaderFields:%@", urlResponse, dataOrLocation, error, httpRes.statusCode, httpRes.allHeaderFields);
        }];
        
        [[BOFNetworkPromiseExecutor sharedInstance] executeNetworkPromise:netpromise];
    } @catch (NSException *exception) {
        BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
        failureBlock_(nil, nil, [NSError boErrorForDict:exception.userInfo]);
    }
}
@end

File: BlotoutAnalytics/Network/BONetworkManager.swift
//
//  BONetworkManager.swift
//  BlotoutAnalytics
//
//  Created by Poonam Tiwari on 19/04/22.
//

import Foundation
//
//public enum BOErrorCodes : Int {
//     case boErrorUnknownn = 10001
//     case boErrorNoInternetConnection
//     case boErrorParsingError
//     case boManifestSyncError
// }

class BONetworkManager:NSObject{
    
    
    
    class func asyncRequest(
        _ request: URLRequest?,
        success successBlock_: @escaping (Any?, URLResponse?) -> Void,
        failure failureBlock_: @escaping (Any?, URLResponse?, Error?) -> Void) {
            
            if BOReachability().currentReachabilityStatus() == .boNotReachable {
                let error = BOErrorAdditions.boError(forCode: BOErrorCodes.boErrorNoInternetConnection.rawValue, withMessage: "Network Not Reachable")
                failureBlock_(nil, nil, error)
                return
            }
            
            //TODO: fix this, always failing
            
            let netpromise = BOFNetworkPromise(urlRequest: request, completionHandler: { urlResponse, dataOrLocation, error in
                let httpRes = urlResponse as? HTTPURLResponse
                if httpRes?.statusCode == 200 {
                    successBlock_(dataOrLocation, urlResponse)
                } else {
                    failureBlock_(dataOrLocation, urlResponse, error)
                }
                BOFLogDebug(frmt: "urlResponse:%@  dataOrLocation:%@ error:%@ urlResponse_statusCode:%ld allHeaderFields:%@", args: urlResponse as! CVarArg, dataOrLocation as! CVarArg , error?.localizedDescription as! CVarArg, httpRes?.statusCode as! CVarArg, httpRes?.allHeaderFields as! CVarArg)
            })
            //                                                   else {
            //                    return failureBlock_(dataOrLocation, urlResponse, error)
            //
            //                }
            
            BOFNetworkPromiseExecutor.sharedInstance.execute(netpromise!)
            
        }
}
